### DISCUSSION TITLE
CI를 통한 유닛/연계 테스트 예시

### DISCUSSION THREAD

**Discussion by jihoon-seo** [cloud-barista/cb-larva#70] [Ideas]
향후 참고/논의/활용 등을 위하여
CI를 통한 유닛/연계 테스트 예시들을 남겨 봅니다.

**Comment by jihoon-seo** [comment]
[[beego/.travis.yml](https://github.com/beego/beego/blob/develop/.travis.yml)]
- Travis CI 의 yaml 파일을 다음과 같이 구성할 수 있네용
  - env
    - global
    - matrix (해당 예시에서는 `sqlite3`, `postgres`, `mysql` 의 3가지 경우를 테스트)
  - before_install
    - [SSDB](https://github.com/ideawu/ssdb) 다운로드 및 빌드 (etcd : nutsdb = Redis : SSDB 정도..?)
    - `etcd` 컨테이너 실행
    - ...
  - install
    - 린터들 (`staticcheck`, `unconvert`, `ineffassign`) 설치
  - before_script
    - SQL DB 초기화 (`create database`)
    - SSDB 서버 실행
  - script
    - `go test -coverprofile=coverage.txt -covermode=atomic ./...`
    - 린터들 (`staticcheck`, `unconvert`, `ineffassign`) 실행
  - after_script
    - SSDB 서버 종료 및 데이터 삭제
  - after_success
    - `codecov.io` 에 결과 전송
  - ...
- 실행 사례
  - https://travis-ci.org/github/beego/beego/builds/756421634
  - https://travis-ci.org/github/beego/beego/builds/756516776
  - https://travis-ci.org/github/beego/beego/builds/756542031
  - https://travis-ci.org/github/beego/beego/builds/756547552

**Comment by yunkon-kim** [comment]
Travis CI 정보공유 감사드립니다. 👍 

GitHub Actions를 통한 Continuous Integration (CI) 시 **SSDB, etcd, codecov 등을 주로 참고하면 좋을 것 같고요.** 😄 

**Linter 종류도 참고하겠습니다.**

추가로, CB-Larva의 Linter 적용상황 공유 드리오니 참고 바랍니다.
- gofmt, govet, gocyclo, golint, ineffassign, misspell를 적용하였고요.
- Go Report Card에서 좋은 평점을 받기 위하여 위 Linter를 우선 적용하였습니다.
- 참고
  - [[.golangci-lint.yml](https://github.com/cloud-barista/cb-larva/blob/master/.golangci.yml)]
  - [![Go Report Card](https://goreportcard.com/badge/github.com/cloud-barista/cb-larva)](https://goreportcard.com/report/github.com/cloud-barista/cb-larva)

**Comment by jihoon-seo** [comment]
[[How To Configure a Continuous Integration Testing Environment with Docker and Docker Compose on Ubuntu 14.04](https://www.digitalocean.com/community/tutorials/how-to-configure-a-continuous-integration-testing-environment-with-docker-and-docker-compose-on-ubuntu-14-04)]
이 외에도 `docker-compose test` 로 검색하면 관련 글들이 있습니다.

---

[[Automated repository tests on Docker Hub](https://docs.docker.com/docker-hub/builds/automated-testing/)]

---

[[위의 내용을 GitHub Actions 로 수행](https://github.com/cloud-barista/cb-spider/blob/master/.github/workflows/docker-publish.yml)]
```YAML
      - name: Run tests
        run: |
          if [ -f docker-compose.test.yml ]; then
            docker-compose --file docker-compose.test.yml build
            docker-compose --file docker-compose.test.yml run sut
          else
            docker build . --file Dockerfile
          fi
```
- 현재는 `docker-compose.test.yml` 가 없어서 `docker build` 만 수행되고 있습니다.
- ...
